core/templates/services/connection-service.service.spec.ts

```diff
@@ -0,0 +1,139 @@
+    },
+    onoffline() {
+      return;
+    }
+  };
+}
+
+describe('Connection Service', () => {
+  let connectionService: ConnectionService;
+  let subscriptions: Subscription;
+  let httpTestingController: HttpTestingController;
+  let connectionStateSpy: jasmine.Spy;
+  let mockWindowRef: MockWindowRef;
+  beforeEach(() => {
+    mockWindowRef = new MockWindowRef();
+    TestBed.configureTestingModule({
+      imports: [
+        HttpClientTestingModule,
+      ],
+      providers: [
+        ConnectionService,
+        {
+          provide: WindowRef,
+          useValue: mockWindowRef
+        }]
+    });
+    connectionService = TestBed.get(ConnectionService);
+    httpTestingController = TestBed.get(HttpTestingController);
+  });
+  beforeEach(() => {
+    connectionStateSpy = jasmine.createSpy('stateChange');
+    subscriptions = new Subscription();
+    subscriptions.add(connectionService.onInternetStateChange.subscribe(
+      connectionStateSpy
+    ));
+  });
+
+  afterEach(() => {
+    subscriptions.unsubscribe();
+    httpTestingController.verify();
+  });
+
+  // Check whether the connection service is initialized correctly.
+  it('should be initialized correctly', () => {
+    expect(connectionService).toBeTruthy();
+  });
+
+  it('should return connection state when offline', () => {
+    connectionService.checkNetworkState();
+    spyOn(connectionService, 'checkNetworkState');
+    mockWindowRef.nativeWindow.onoffline();
+    expect(connectionStateSpy).toHaveBeenCalledWith({
+      hasNetworkConnection: false,
+      hasInternetAccess: false
+    });
+  });
+
+  it('should return connection state when online', () => {
+    connectionService.checkNetworkState();
+    spyOn(connectionService, 'checkNetworkState');
+    mockWindowRef.nativeWindow.ononline();
+    expect(connectionStateSpy).toHaveBeenCalledWith({
+      hasNetworkConnection: true,
+      hasInternetAccess: true
+    });
+  });
+
+  it('should return online when is connected to internet', fakeAsync(() => {
+    connectionService.checkInternetState();
+    tick(100);
+    discardPeriodicTasks();
+    let req = httpTestingController.expectOne('/connectivity/check');
+    expect(req.request.method).toEqual('GET');
+    req.flush({
+      isInternetConnected: true,
+    });
+    flushMicrotasks();
+    expect(connectionStateSpy).toHaveBeenCalledWith({
+      hasNetworkConnection: true,
+      hasInternetAccess: true
+    });
+  }));
+
+  it('should return online when reconnected after offline', fakeAsync(() => {
+    connectionService.checkInternetState();
+    connectionService.checkNetworkState();
+    tick(100);
+    let req = httpTestingController.expectOne('/connectivity/check');
+    expect(req.request.method).toEqual('GET');
+    req.flush({
+      isInternetConnected: true,
+    });
+    flushMicrotasks();
+    mockWindowRef.nativeWindow.onoffline();
+    tick(5000);
+    mockWindowRef.nativeWindow.ononline();
+    tick(8000);
+    discardPeriodicTasks();
```

Comment:
Can you please add inline comments here to describe the test steps please so that it is clearer?

Reply:
Done.